### DISCUSSION TITLE
DevTools in production from PR #44

### DISCUSSION THREAD

**Discussion by arjunvegda** [jotaijs/jotai-devtools#53] [General]
#44 
Reference: https://github.com/jotaijs/jotai-devtools/pull/44#issuecomment-1494543198

**Comment by arjunvegda** [comment]
@sari-shipt - What version of NextJS, are you using? I'll try to reproduce it on my end. Also, I believe @blanky0230 is using CRA and not NextJS.

**Reply by arjunvegda** [reply]
Also, just to triple check — [Step 5](https://github.com/jotaijs/jotai-devtools/pull/44#issuecomment-1474866468) from [this comment](https://github.com/jotaijs/jotai-devtools/pull/44#issuecomment-1474866468) isn't working for you either, right?

**Reply by sari-shipt** [reply]
It ended up being something off within our config. I tried all options suggested last week, but after modifying the config further and using Step 5 our build worked as expected. Thanks for your time!

**Reply by arjunvegda** [reply]
Hey @sari-shipt @blanky0230

Could you help us test this build? We're expecting tree-shaking to automatically work without needing to wrap the component with `process.env` on userland.

```shell
yarn add jotai-devtools@https://pkg.csb.dev/jotaijs/jotai-devtools/commit/2b2603e6/jotai-devtools/_pkg.tgz
```

**Example**
```tsx
import { DevTools } from 'jotai-devtools'

const App = () => {
  return (
    <>
      <DevTools/> // <-- This should automatically be tree-shaken on production builds
    </>
  );
};
```

**Reply by SariSabouh** [reply]
Sure. Testing!

**Reply by SariSabouh** [reply]
Build went from `614kb` to `1.46mb` after replacing removing the `process.env` check from:

```
process.env.NODE_ENV === 'development' ? (
    <JotaiDevTools store={store} /> : null
```

**Reply by arjunvegda** [reply]
Thanks for checking! That's odd, it should be tree-shaken out.

Are you testing on NextJs? 🤔

**Reply by SariSabouh** [reply]
Yes. I pretty much removed the .env and ran build with ANALYZE=true

**Reply by arjunvegda** [reply]
Interesting! Could you help us reproduce it? 

NextJS build caches the node modules within `.next` so could you try running `rm -rf .next && next build` and see if that makes any difference?

I created a [CodeSandbox](https://codesandbox.io/p/sandbox/jotai-devtools-discussions-52-njt98n) and it works as expected. Hit "+" > "Build" to trigger the build 

### Before
<img width="504" alt="image" src="https://user-images.githubusercontent.com/14841132/234432618-c1d9367f-d9d7-435d-b8c3-0de2df8189e9.png">

### After
<img width="499" alt="image" src="https://user-images.githubusercontent.com/14841132/234435896-45cde767-64ce-4c17-8921-9c4b8e22356a.png">

**Reply by SariSabouh** [reply]
Just tested it and it didn't work, @arjunvegda . Let me know if there is anything else you'd like me to try.

**Reply by arjunvegda** [reply]
Could you help us create a small repro in that case? Feel free to fork [my CodeSandbox above](https://codesandbox.io/p/sandbox/jotai-devtools-discussions-52-njt98n) ^

**Reply by SariSabouh** [reply]
The crazy part is that I did not face this issue either whenever I did it on CodeSandbox 🤔 ....

**Reply by arjunvegda** [reply]
Oh, so it's local only? Feel free to push it to a GitHub repo then

**Reply by SariSabouh** [reply]
It's happening to my company's project and it's a private repo unfortunately. So I can't share it. The CodeSandbox quick demo project I created worked just fine with the tree-shaking.

**Reply by SariSabouh** [reply]
The issue persists in my company's project without the `process.env`

**Comment by arjunvegda** [comment]
@dai-shi - What are your thoughts on adding a backward compatible check for `process.env.` to [ESM modules](https://github.com/jotaijs/jotai-devtools/blob/main/tsup.config.ts#L58) until the industry fully adopts `import.meta.env` 🤔 

Perhaps we could ship something like this? 
```ts
((import.meta.env&&import.meta.env.MODE)!=="production" || (process?.env?.NODE_ENV)!=="production"))
```

**Reply by dai-shi** [reply]
well, that's very tricky and `process?` doesn't work.
is this only about jotai-devtools? or also jotai? (I kind of am aware import.meta.env.MODE doesn't work even with vite. So, I'm happy to think about some intermediate solutions. Maybe, with more complex export conditions)

**Reply by arjunvegda** [reply]
> well, that's very tricky and process? doesn't work.

Ah, right! When you say "with more complex export conditions", do you mean exporting multiple variants of components each equipped with `import.meta...` and `process.env...`?

> is this only about jotai-devtools? or also jotai?

Appears to be both actually 🤔. I verified it locally using CRA. Those `debugPrivate` labels from `jotai` are still being added when creating a prod build and wrapping them with `process.env.NODE_ENV` fixes it.

> import.meta.env.MODE doesn't work even with vite

We just got another [report](https://github.com/jotaijs/jotai-devtools/issues/54) that confirms this too

**Reply by dai-shi** [reply]
For vite, we need ask them to replace `import.meta.env` too. Do we have issues other than vite? We would need to configure "export conditions" for each case. `import.meta` is only available for ESM, and `process` is only available for node and some bundlers.

**Reply by arjunvegda** [reply]
>  Do we have issues other than vite?

We also got reports for Next JS and Create React App (mostly Webpack?) too.

Do most bundlers support `production` and/or `development` conditions? Could we use those?

Any thoughts on separate builds for browser and node? Here is a rough example

```json
{
   "exports":{
      ".":{
         "node":{
            "import":{
               "types":"./dist/index.d.ts",
               "default":"./dist/index.cjs.js"
            }
         },
         "browser":{
            "import":{
               "types":"./dist/index.d.mts",
               "default":"./dist/index.esm.mjs"
            }
         },
         "types":"./dist/index.d.ts",
         "default":"./dist/index.cjs.js"
      }
   }
}
```

**Reply by dai-shi** [reply]
Yeah, I meant separate exports like that.
There are a few issues:
- New node should use ESM instead of CJS
- However, it's not common to use `import.meta.env.MODE` in Node, it's just Vite convention.
- But, even Vite can't handle `(import.meta.env && import.meta.env.MODE) !== "production"` as of now.
- Some bundlers (Webpack?) pick ESM, but they actually don't support ESM and transpile to CJS (import.meta causes an error).

So, unless we give up following Vite convention, we should provide ESM with `process.env.NODE_ENV` for this migration period.

We actually did so in v1 with `"module"` condition. (We still do in Zustand and Valtio.)
So, if we find good combinations of conditions, we can use it instead of `"module"`.
